src/ripple/app/tx/impl/URIToken.cpp

```diff
@@ -0,0 +1,819 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2017 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <ripple/app/ledger/Ledger.h>
+#include <ripple/app/tx/impl/URIToken.h>
+#include <ripple/basics/Log.h>
+#include <ripple/protocol/Feature.h>
+#include <ripple/protocol/Indexes.h>
+#include <ripple/protocol/Quality.h>
+#include <ripple/protocol/STAccount.h>
+#include <ripple/protocol/TER.h>
+#include <ripple/protocol/TxFlags.h>
+
+namespace ripple {
+
+NotTEC
+URIToken::preflight(PreflightContext const& ctx)
+{
+    if (!ctx.rules.enabled(featureURIToken))
+        return temDISABLED;
+
+    NotTEC const ret{preflight1(ctx)};
+    if (!isTesSuccess(ret))
+        return ret;
+
+    uint32_t flags = ctx.tx.getFlags();
+    uint16_t tt = ctx.tx.getFieldU16(sfTransactionType);
+
+    switch (tt)
+    {
+        case ttURITOKEN_MINT: {
+            if (flags & tfURITokenMintMask)
```

Comment:
what do you think about splitting out the bodies of each switch statement into their own functions, for readability?

Reply:
answered below in followup question.